### PR TITLE
Initial support for Kubernetes 1.12+

### PR DESCRIPTION
**Pull request by outcoldman**
Fixes https://github.com/kubernetes/minikube/issues/3164

Support for k8s 1.12+, also keeping support for old versions of k8s.

See this comment for the list of incompatibilities with 1.12 https://github.com/kubernetes/minikube/issues/3164#issuecomment-425327164

1. It uses `kubeadm.k8s.io/v1alpha1` for kubeadm. Kubernetes 1.12 does not support version `v1alpha1`. Only `v1alpha2`. The recommended version is `v1alpha3`. I used `kubeadm config  upgrade ...` to convert v1alpha1 to v1alpha1 and to v1alpha3.

2. This addon is not available `/usr/bin/kubeadm alpha phase addon kube-dns`, I guess it should use `coredns` instead (change in kubeadm to change `kube-dns` to `coredns `)

3. The argument `cadvisor-port` is not available in k8s-1.12

### CODE REVIEW

**Comment by dlorenc** [comment]
@minikube-bot OK to test

**Comment by outcoldman** [comment]
@dlorenc is it related to my change? or just a test issue? I can change back ‘1.10’ as default, that probably should keep all the test green.

**Comment by outcoldman** [comment]
@dlorenc I switched it back to `1.10.0` as default, that way the tests should work.

**Comment by outcoldman** [comment]
tests are still failing, is it me or some configuration issues?

**Comment by tamalsaha** [comment]
@outcoldman , please see https://github.com/kubernetes/minikube/pull/3132#issuecomment-423725400 comment. I am not sure if that is relevant.

**Comment by outcoldman** [comment]
@tamalsaha hard to tell, considering that I am using by default the same version as before, not sure how that can affect tests

**Comment by tstromberg** [comment]
@minikube-bot OK to test

File: pkg/minikube/bootstrapper/kubeadm/kubeadm.go [hunk 1]
```diff
@@ -119,12 +119,17 @@ func (k *KubeadmBootstrapper) StartCluster(k8s config.KubernetesConfig) error {
 		KubeadmConfigFile   string
 		SkipPreflightChecks bool
 		Preflights          []string
+		DNSAddon            string
```
Comment (tstromberg) [inline pkg/minikube/bootstrapper/kubeadm/kubeadm.go]: Can you elaborate in the PR description or commit message on why this change is necessary?

kube-Perhaps something about the default changing from kube-dns to coredns in 1.12?

File: pkg/minikube/bootstrapper/kubeadm/templates.go [hunk 2]
```diff
@@ -44,6 +44,45 @@ nodeName: {{.NodeName}}
...
+var kubeadmConfigTemplateV1Alpha3 = template.Must(template.New("kubeadmConfigTemplate-v1alpha3").Funcs(template.FuncMap{
+	"printMapInOrder": printMapInOrder,
+}).Parse(`apiEndpoint:
+  advertiseAddress: {{.AdvertiseAddress}}
+  bindPort: {{.APIServerPort}}
+apiVersion: kubeadm.k8s.io/v1alpha3
+bootstrapTokens:
+- groups:
+  - system:bootstrappers:kubeadm:default-node-token
+  token: 599s02.8i3w2yyyay8t0onm
```
Comment (tstromberg) [inline pkg/minikube/bootstrapper/kubeadm/templates.go]: How was this value arrived at?

File: pkg/minikube/bootstrapper/kubeadm/versions.go [hunk 3]
```diff
@@ -260,6 +259,14 @@ var versionSpecificOpts = []VersionedExtraOption{
 		},
 		GreaterThanOrEqual: semver.MustParse("1.11.0-alpha.0"),
 	},
+	{
+		Option: util.ExtraOption{
```
Comment (tstromberg) [inline pkg/minikube/bootstrapper/kubeadm/versions.go]: Mention this change as part of the PR description.

**Comment by tstromberg** [comment]
FYI: The remaining flakes appear to be due to #3163 - now resolved. There are other flakes, of course.

@minikube-bot OK to test

**Comment by AkihiroSuda** [comment]
Please update Docker to v18.06.1 for supporting `ProcMount`: https://github.com/kubernetes/kubernetes/pull/64283 https://github.com/docker/docker-ce/commit/67fe100ad5e784c8d273e4de768974ecc31a8d59#diff-04945848696f0f99c75256e2edab8111

File: pkg/minikube/bootstrapper/kubeadm/templates.go (hunk 2 above)
Comment (outcoldman) [reply pkg/minikube/bootstrapper/kubeadm/templates.go]: I have used `kubeadm config upgrade`, cleared all the default values, just by accident left this one. I will remove it.

File: pkg/minikube/bootstrapper/kubeadm/versions.go (hunk 3 above)
Comment (outcoldman) [reply pkg/minikube/bootstrapper/kubeadm/versions.go]: Updated PR description. I had a link to comment https://github.com/kubernetes/minikube/issues/3164#issuecomment-425327164 that explains the reasons, but agree to have it in PR is better.

File: pkg/minikube/bootstrapper/kubeadm/kubeadm.go (hunk 1 above)
Comment (outcoldman) [reply pkg/minikube/bootstrapper/kubeadm/kubeadm.go]: Updated PR description. I had a link to comment #3164 (comment) that explains the reasons, but agree to have it in PR is better.

**Comment by outcoldman** [comment]
@tstromberg update PR description and removed token generated by `kubeadm config upgrade`
